### PR TITLE
improve sub-second datetime plotting and documentation

### PR DESCRIPTION
**Pull request by nmartensen**
## PR Summary

Users trying to plot extremely short (sub-millisecond) time intervals using datetime will run into issues, as shown in #10073.  I'd like to make the limitations more explicit in the documentation, and investigate if the situation can be improved without making things much more complicated.

Feedback is welcome.

Main changes right now are:
- the addition of a note to the MicrosecondLocator docstring, and
- a small change to microsecond rounding.

The rounding change improves the situation when using the early-year workaround and should not have any effect anywhere else.

The remaining issue (the reason for the warning at the end of the note) is the 10-millisecond "snapping distance" around full seconds (dates.py lines 292 to 296). This is needed in order to produce correct tick labels for plots of longer time intervals, therefore it cannot simply be dropped.

Perhaps an optional `microsecond_precision` parameter (default: 20) could be added to `_from_ordinalf()` and `num2date()`, with microseconds then being rounded to nearest integer multiple of this?  The MicrosecondLocator would be the only one to use it, and can set it to 1. Would something like this be acceptable?  Then the warning in the note could be dropped.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by nmartensen** [comment]
updated with new commits:
- replace the original full-second snapping with musec-interval rounding
- DateFormatter: use full musec precision tick labeling when using '%f' in strftime format string, but otherwise always round
- add a warning when detecting years where the precision might not be sufficient.

**Comment by nmartensen** [comment]
With these commits, output with year <= 20 now looks exactly like the "Expected outcome" in #10073. Higher years trigger a warning. The precision issues with modern years are clearly visible in the tick labels and are not hidden in any way.

**Comment by nmartensen** [comment]
more updates:
- commit messages provide more background information
- code changes are PEP 8 compliant
- updated test results (test_axes.py::test_date_timezone_\*, test_dates.py::test_DateFormatter, test_auto_date_locator\*)

Unfortunately github does not support stacked pull requests yet. Otherwise this WIP would consist of three PRs, one on top of the next:
1. documentation changes (first three commits)
2. changes to microsecond rounding (remaining commits)
3. improved millisecond tick labeling without microseconds (not included yet)

Commits implementing (3) already exist, but that's probably too much for this single PR. If you prefer, (1) and (2) can also be split into separate pull requests.

**Comment by tacaswell** [comment]
@nmartensen Thanks for working on this!

However, I am not sure that this is the best approach here.  Adding a kwarg to all of the time related functions is a bit disruptive. 

How hard would it be to up the precision of the internal storage type?

**Comment by efiring** [comment]
It seems to me that time-handling with high precision is so fundamentally broken (e.g., it ignores leap seconds) with all calendar-related code, not to mention matplotlib's use of double precision days since an absurdly long-ago epoch, that trying to patch it up with a hack--a fake year--is just making things more confusing.  Wouldn't it make more sense for the user to pick an origin (e.g., the beginning of an experiment) and label the axis with microseconds (or milliseconds, or floating-point seconds) from that origin?  Maybe make custom locators and formatters?

**Comment by jklymak** [comment]
`datetime` supports microseconds, and we purport to plot `datetimes`. So I guess we should be able to handle microsecond input.  

1. We could stop translating to epoch and let the  unit converter carry around a time=0 internal variable (probably easiest if its a day=0 datetime).  
2. We could make datenums be `np.longdouble`.

I think 2 is easiest, though I'm not sure if our plotting respects `np.longdouble?` which you'd want to get the actual position on the axis correct.   

I think 1 is most platform independent?    1) would probably just require `date2num` and `num2date` to have optional `day=0` kwargs that allow an offset in calculating the floating-point value for drawing.

**Comment by efiring** [comment]
Using long double is not an option--essentially, there is no such thing. See https://docs.scipy.org/doc/numpy-dev/user/basics.types.html.

Changing our datenum is a *big* api break.  If we add an alternative--like datetime64, in one or more of its variants--we would have to maintain full compatibility with existing datenums.  This might be the way to go, since we should have native datetime64 support anyway.

What we have is a bit silly, but it works for normal dates and times.  Datetime *has* microseconds, but given that it doesn't handle leap seconds, it is a mistake to think it is *handling* time units with high accuracy.  Furthermore, what is a sensible way of labeling ticks on a time axis covering extremely short intervals?  I doubt that it involves year, month, day, hour, minute, second, and microseconds.

I think the best course of action here is to start by documenting the present limitation, and then carefully think through strategies for improvement, starting with the need for our own datetime64 support (recognizing that datetime64 is flawed, but we are probably stuck with it.)

**Comment by jklymak** [comment]
@efiring yeah I agree. 

If thats the consensus, maybe it makes more sense to deprecate the microsecond Locator so people don't expect it to work?  

WRT `datetime64`, we already have support: #9779

**Comment by efiring** [comment]
On 2018/01/02 3:00 PM, Jody Klymak wrote:
> @efiring <https://github.com/efiring> yeah I agree.
> 
> If thats the consensus, maybe it makes more sense to deprecate the 
> microsecond Locator so people don't expect it to work?

That makes sense to me.

> 
> WRT |datetime64|, we already have support: #9779 
> <https://github.com/matplotlib/matplotlib/pull/9779>

Great! I completely forgot about it.

**Comment by nmartensen** [comment]
> If thats the consensus, maybe it makes more sense to deprecate the microsecond Locator so people don't expect it to work?

Users not knowing about implementation details (such as microsecondlocators) may still expect it to work, since plotting datetime.datetime in general is supported, and datetime.datetime has microseconds.

What's wrong with documenting the limitations and making it work as good as possible? The note added to the documentation already contains advice to use some other time format.

**Comment by jklymak** [comment]
If it could be done without a lot of extra code outside the locator and formatter it’s probably fine. 

I think the current situation plus a better doc/info output would suffice.

File: lib/matplotlib/dates.py [hunk 1]
```diff
@@ -1296,6 +1304,11 @@ def get_locator(self, dmin, dmax):
             locator = RRuleLocator(rrule, self.tz)
         else:
             locator = MicrosecondLocator(interval, tz=self.tz)
+            if dmin.year > 20 and interval < 1000:
+                warnings.warn('Plotting microsecond time intervals is not'
```
Comment (jklymak) [inline lib/matplotlib/dates.py]: I'd do `_log.warn` so knowledgeable users can turn this warning off.

**Comment by efiring** [comment]
What is the use case for a microsecond locator?  Under what circumstances does it really make sense to use datetime machinery and calendars for generating tick labels for extremely short intervals?  What can a user do with such a locator that cannot be done as well or better with plain numbers, using non-date-specific locators?
Sometimes things get into mpl before we realize they shouldn't; I think the microsecond locator is in that category.  Matplotlib suffers from bloat.  When we have the opportunity to trim it out, we should do so.
I'm open to being convinced otherwise if there is a compelling use case.

File: lib/matplotlib/dates.py [hunk 2]
```diff
@@ -286,14 +286,22 @@ def _from_ordinalf(x, tz=None):
 
     remainder = float(x) - ix
 
-    # Round down to the nearest microsecond.
-    dt += datetime.timedelta(microseconds=int(remainder * MUSECONDS_PER_DAY))
+    # Round to the nearest microsecond.
+    dt += datetime.timedelta(
```
Comment (jklymak) [inline lib/matplotlib/dates.py]: This could be an `else` of the `if` below, couldn't it?

File: lib/matplotlib/dates.py [hunk 3]
```diff
@@ -96,11 +96,11 @@
       <../gallery/ticks_and_spines/date_demo_rrule.html>`_.
 
     * :class:`AutoDateLocator`: On autoscale, this class picks the best
-      :class:`RRuleLocator` to set the view limits and the tick
+      :class:`DateLocator` to set the view limits and the tick
```
Comment (jklymak) [inline lib/matplotlib/dates.py]: I think in general its more useful to point to the `RRuleLocator` than the generic `DateLocator`.  If you want to be pedantic, you can refer to both.

Comment (nmartensen) [reply lib/matplotlib/dates.py]: In my first draft I had been referring to both, but then found it too complicated/pedantic for this description. The ``Yearlocator`` is also no ``RRuleLocator``.

File: lib/matplotlib/dates.py (hunk 2 above)
Comment (nmartensen) [reply lib/matplotlib/dates.py]: Then the code below would need to change. The above is needed to set the hours, minutes, and seconds; the stuff below changes only microseconds.

**Comment by jklymak** [comment]
@efiring I think this PR improves the situation.  It actually simplifies the microsecond handling, and it improves the documentation and adds a warning to the user that MPL/datetime is doing something imprecise.  We can talk more about deprecation later, as better Datetime handling is still a todo item I have so it won't be forgotten.

**Comment by jklymak** [comment]
@nmartensen is this still a WIP?

**Comment by nmartensen** [comment]
No, you are right. I don't have plans for further changes, unless you have more comments. Changed the title accordingly.

**Comment by jklymak** [comment]
Great - it needs a second approver before we can merge.  @efiring appears somewhat against, so that may squash other support :wink: which means it *might* not get merged.  This happens sometimes. However, your contribution is very much aprpeciated, and if it doesn't get included, please *do* take advantage of having the machinery set up to send more PRs.

**Comment by jklymak** [comment]
Thanks @efiring  and thanks a lot @nmartensen !

**Pull request merged**